app/models/manageiq/providers/amazon/cloud_manager/ssa_agent_manager.rb

```diff
@@ -0,0 +1,326 @@
+    return false unless bucket.exists?
+
+    obj_id = heartbeat_prefix + agent_id
+    obj = bucket.object(obj_id)
+    return false unless obj.exists?
+   
+    last_beat_stamp = YAML.load(obj.get.body.read, safe: true)
+    Time.now.utc - last_beat_stamp > interval ? false : true
+  end
+
+  def deploy_agent
+    kp = get_key_pair
+    security_group_id = create_security_group
+    data = create_user_data
+    zone_name = ec2.client.describe_availability_zones.availability_zones[0].zone_name
+    subnets = get_subnets(zone_name)
+    raise "No subnet_id is available for #{zone_name}!" if subnets.length == 0
+    create_profile
+
+    instance = ec2.create_instances({
+      image_id: get_agent_image_id,
+      min_count: 1,
+      max_count: 1,
+      key_name: kp.name,
+      security_group_ids: [security_group_id],
+      user_data: data,
+      instance_type: 't2.micro',
+      placement: {
+        availability_zone: zone_name
+      },
+      subnet_id: subnets[0].subnet_id,
+      iam_instance_profile: {
+        name: profile_name
+      },
+      tag_specifications: [{
+        resource_type: "instance",
+        tags: [{
+          key: "Name",
+          value: MIQ_SSA
+        }]
+      }]
+    })
+    ec2.client.wait_until(:instance_status_ok, {instance_ids: [instance[0].id]})
+
+    instance[0].id
+  end
+
+  private
+  def role_exists?(role_name)
+    begin
+      role = iam.role(role_name)
+      role.role_id
+      true
+    rescue ::Aws::IAM::Errors::NoSuchEntity
+      false
+    end
+  end
+
+  def create_role(role_name = MIQ_SSA)
+    return if role_exists?(role_name)
+
+    # Policy Generator:
+    policy_doc = {
+      "Version": "2012-10-17",
+      "Statement": [
+        {
+          "Effect": "Allow",
+          "Principal": {"Service": "ec2.amazonaws.com"},
+          "Action": "sts:AssumeRole"
+        }
+      ]
+    }
+
+    role = iam.create_role({
+      role_name: role_name,
+      assume_role_policy_document: policy_doc.to_json
+    })
+    
+    # grant all priviledges
+    role.attach_policy({
+      policy_arn: 'arn:aws:iam::aws:policy/AmazonS3FullAccess'
+    })
+    
+    role.attach_policy({
+      policy_arn: 'arn:aws:iam::aws:policy/AmazonEC2FullAccess'
+    })
+    
+    role.attach_policy({
+      policy_arn: 'arn:aws:iam::aws:policy/AmazonSQSFullAccess'
+    })
+
+    role
+  end
+
+  def create_profile(profile_name = MIQ_SSA, role_name = MIQ_SSA)
+    ssa_profile = iam.instance_profile(profile_name)
+    ssa_profile = iam.create_instance_profile(instance_profile_name: profile_name) unless ssa_profile.exists?
+
+    create_role(role_name) unless role_exists?(role_name)
+    ssa_profile.add_role(role_name: role_name) if ssa_profile.roles.size == 0
+
+    ssa_profile
+  end
+
+  def get_agent_image_id
+    imgs = ec2.client.describe_images(
+      filters: [
+        {
+          name: "name",
+          values: [ "RHEL-7.3_HVM_GA*" ]
+        }
+      ]
+    ).images
+
+    imgs[0].image_id
```

Comment:
Is the above value always correct?  Do we know that the first image with that prefix will always be the one we want?